### PR TITLE
ipc - support for `Uint8Array`

### LINKED ISSUE #154820: Explorer: cannot copy paste anymore to duplicate file with sandbox enabled
* enable `window.experimental.useSandbox`
* copy & paste from explorer

=> 🐛 

```
The first argument must be of type string or an instance of Buffer, ArrayBuffer, or Array or an Array-like Object. Received an instance of Object
```

### PR DESCRIPTION
**Pull request by bpasero**
cc @alexdima this enables `Uint8Array` support for IPC. We need this because when `sandbox: true` all `VSBuffer.buffer` will be `Uint8Array` and not `Buffer` anymore.

This PR fixes https://github.com/microsoft/vscode/issues/154820

### CODE REVIEW

**Review by joaomoreno — COMMENTED** [review]
Why not also add support for all other typed arrays? They all work the same way: there's an underlying byte buffer which we can take literally, we just need to remember to instantiate the right typed array on the receiving side.

File: src/vs/base/parts/ipc/common/ipc.ts [hunk 1]
```diff
@@ -265,8 +272,10 @@ function deserialize(reader: IReader): any {
 	switch (type) {
 		case DataType.Undefined: return undefined;
 		case DataType.String: return reader.read(readSizeBuffer(reader)).toString();
-		case DataType.Buffer: return reader.read(readSizeBuffer(reader)).buffer;
 		case DataType.VSBuffer: return reader.read(readSizeBuffer(reader));
+		case DataType.Uint8Array:
+		case DataType.Buffer:
+			return reader.read(readSizeBuffer(reader)).buffer;
```
[the remark this answers is not in the document]
Comment (bpasero) [reply src/vs/base/parts/ipc/common/ipc.ts:278]: To clarify, you would want a way for `VSBuffer` to return a new `Uint8Array` and not the underlying node.js `Buffer`? Why can we not return the node.js `Buffer`, it extends `Uint8Array` after all:

https://github.com/microsoft/vscode/pull/154916/commits/7325b7a9af7e3c582a157b1e724bc693bcc4145e

File: src/vs/base/parts/ipc/test/common/ipc.test.ts [hunk 2]
```diff
@@ -319,6 +331,11 @@ suite('Base IPC', function () {
 			const r = await ipcService.buffersLength([VSBuffer.alloc(2), VSBuffer.alloc(3)]);
 			return assert.strictEqual(r, 5);
 		});
+
+		test('uint8arrays in arrays', async function () {
+			const r = await ipcService.uint8ArrayLength([VSBuffer.alloc(2).buffer, VSBuffer.alloc(3).buffer]);
+			return assert.strictEqual(r, 5);
```
[the remark this answers is not in the document]
Comment (bpasero) [reply src/vs/base/parts/ipc/test/common/ipc.test.ts:337]: I do here: https://github.com/microsoft/vscode/blob/7325b7a9af7e3c582a157b1e724bc693bcc4145e/src/vs/base/parts/ipc/test/common/ipc.test.ts#L144

Where else would you assert that?

**Comment by bpasero** [comment]
> Why not also add support for all other typed arrays? They all work the same way: there's an underlying byte buffer which we can take literally, we just need to remember to instantiate the right typed array on the receiving side.

But then we need to have a `DataType` per kind of typed array here?

https://github.com/microsoft/vscode/blob/7325b7a9af7e3c582a157b1e724bc693bcc4145e/src/vs/base/parts/ipc/common/ipc.ts#L273

For now I went with `Uint8Array` only because that is typically our way of data transfer between processes, I did not came across the others for IPC needs. Also, it has the nice advantage of node.js `Buffer` being compatible right away.

**Comment by joaomoreno** [comment]
> But then we need to have a DataType per kind of typed array here?

Yeah we'd record them as different data types. They are all `Buffer` underneath the cover. But remember that `Buffer` doesn't exist on browser, just `ArrayBuffer`.

**Comment by bpasero** [comment]
Maybe its easier you add it as you see fit, I am not sure I get it. We would have to add a conversion for each of these?

image

**Comment by bpasero** [comment]
Better solution in https://github.com/microsoft/vscode/pull/155044

**Pull request closed without merging**